src/main/java/com/example/bookstorewebapp/service/shoppingcart/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,81 @@
+package com.example.bookstorewebapp.service.shoppingcart.impl;
+
+import com.example.bookstorewebapp.dto.cartitem.CreateCartItemRequestDto;
+import com.example.bookstorewebapp.dto.cartitem.UpdateCartItemRequestDto;
+import com.example.bookstorewebapp.dto.shoppingcart.ShoppingCartResponseDto;
+import com.example.bookstorewebapp.exception.EntityNotFoundException;
+import com.example.bookstorewebapp.mapper.CartItemMapper;
+import com.example.bookstorewebapp.mapper.ShoppingCartMapper;
+import com.example.bookstorewebapp.model.CartItem;
+import com.example.bookstorewebapp.model.ShoppingCart;
+import com.example.bookstorewebapp.model.User;
+import com.example.bookstorewebapp.repository.shoppingcart.ShoppingCartRepository;
+import com.example.bookstorewebapp.service.book.BookService;
+import com.example.bookstorewebapp.service.cartitem.CartItemService;
+import com.example.bookstorewebapp.service.shoppingcart.ShoppingCartService;
+import java.util.List;
+import java.util.Objects;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@RequiredArgsConstructor
+@Service
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper cartMapper;
+    private final CartItemService cartItemService;
+    private final CartItemMapper cartItemMapper;
+    private final BookService bookService;
+
+    @Override
+    public ShoppingCart createShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        return shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto getByUserId(Long id) {
+        ShoppingCart shoppingCart = shoppingCartRepository
+                .findShoppingCartByUserId(id).orElseThrow(
+                        () -> new EntityNotFoundException("There is no cart with id: " + id)
+                );
+        return cartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public void update(Long id, UpdateCartItemRequestDto requestDto) {
+        cartItemService.updateById(id, requestDto.getQuantity());
+    }
+
+    @Override
+    public void save(User user, CreateCartItemRequestDto requestDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findShoppingCartByUserId(user.getId())
+                .orElse(createShoppingCart(user));
+        bookService.isExist(requestDto.getBookId());
+        CartItem cartItem = cartItemMapper.toModel(requestDto);
+        cartItem.setShoppingCart(shoppingCart);
+        saveCartItem(requestDto, shoppingCart, cartItem);
+    }
+
+    private void saveCartItem(
```

Comment:
I'd just avoid doing some custom and implicit stuff like that, and would have thrown exception if we already have such book in the cart